torch/distributed/device_mesh.py

```diff
@@ -392,81 +95,123 @@ def _get_device_handle(device_type: str = "cuda"):
+    # https://github.com/NVIDIA/cutlass/blob/6dd13d42784ee5bfa232d2441e6b9a021c5c6290/python/pycute/layout.py#L137
+    def coalesce(l: Layout) -> Layout:
+        res = Layout([])
+        for size, stride in l.sizes_and_strides:
+            if size == 1:
+                pass
+            elif res.sizes_and_strides and stride == res.sizes[-1] * res.strides[-1]:
+                prev_size, prev_stride = res.sizes_and_strides.pop()
+                res.sizes_and_strides.append((size * prev_size, prev_stride))
+            else:
+                res.sizes_and_strides.append((size, stride))
+        return res
+
+    # Borrowed from pycute.layout.composition
+    # https://github.com/NVIDIA/cutlass/blob/6dd13d42784ee5bfa232d2441e6b9a021c5c6290/python/pycute/layout.py#L190
+    def composition(l: Layout, sub_sizes: list[int]) -> list[Layout]:
+        res: list[Layout] = []
+        numel_so_far = 1
+        for sub_size in sub_sizes:
+            sub_stride = numel_so_far
+            numel_so_far *= sub_size
+
+            sub_res = Layout([])
+
+            for curr_size, curr_stride in l.sizes_and_strides[:-1]:
+                assert curr_size % sub_stride == 0 or sub_stride % curr_size == 0
+                new_size = min(max(1, curr_size // sub_stride), sub_size)
+
+                if new_size != 1:
+                    sub_res.sizes_and_strides.append(
+                        (new_size, sub_stride * curr_stride)
+                    )
 
+                sub_size = sub_size // new_size
+                sub_stride = ceil_div(sub_stride, curr_size)
+
+            if sub_size != 1 or len(sub_res.sizes_and_strides) == 0:
+                sub_res.sizes_and_strides.append((sub_size, sub_stride * l.strides[-1]))
+
+            res.append(sub_res)
+
+        return res
+
+    # Borrowed from pycute.layout.complement
+    # https://github.com/NVIDIA/cutlass/blob/6dd13d42784ee5bfa232d2441e6b9a021c5c6290/python/pycute/layout.py#L232
+    def complement(l: Layout) -> Layout:
+        res = Layout([])
+        current_idx = 1
+
+        for size, stride in sorted(l.sizes_and_strides, key=lambda x: x[1]):
+            if stride == 0 or size == 1:
+                continue
+
+            assert current_idx <= size * stride
+
+            res.sizes_and_strides.append((stride // current_idx, current_idx))
+            current_idx = size * stride
+
+        res.sizes_and_strides.append(
+            (ceil_div(get_world_size(), current_idx), current_idx)
+        )
+
+        return coalesce(res)
+
+    def rank_tensor_to_layouts(mesh: torch.Tensor) -> list[Layout]:
+        return [Layout([(mesh.size(i), mesh.stride(i))]) for i in range(mesh.ndim)]
+
+    def layout_to_rank_group_0(l: Layout) -> list[int]:
+        return [
+            sum(c * s for c, s in zip(coord, l.strides))
+            for coord in product(*(range(s) for s in l.sizes))
+        ]
+
+    def layout_to_all_rank_groups(l: Layout) -> list[list[int]]:
```

Comment:
maybe I am missing something, why this is only 2D? Previously `pg_ranks_by_dim` can be any dimension.